programs/uxd/src/mercurial_utils/calculate_possible_lp_token_precision_loss_collateral_value.rs

```diff
@@ -0,0 +1,37 @@
+use crate::error::UxdError;
+use anchor_lang::{
+    prelude::{Account, Clock, SolanaSysvar},
+    Result,
+};
+use fixed::types::I80F48;
+use mercurial_vault::state::Vault;
+
+// Calculate how much collateral could be lost in possible LP token precision loss
+pub fn calculate_possible_lp_token_precision_loss_collateral_value(
+    mercurial_vault: &Account<Vault>,
+    mercurial_vault_lp_mint_supply: u64,
+) -> Result<u64> {
+    let current_time = u64::try_from(Clock::get()?.unix_timestamp)
+        .ok()
+        .ok_or(UxdError::MathError)?;
+
+    // Calculate the price of 1 native LP token
+    // Do not use mercurial_vault.get_amount_by_share because it does not handle precision loss
+    let total_amount = mercurial_vault
+        .get_unlocked_amount(current_time)
+        .ok_or(UxdError::MathError)?;
+
+    let one_lp_token_collateral_value = I80F48::from_num(1)
```

Comment:
same comment as before, I think this particular piece of math does not require the use of I80F48, this would be much simpler as a simple chain of checked_mul/checked_div

Reply:
The maths looks good tho

Reply:
note: I imagine it was written I80F48 because of the div+ceil combo, but that can be achieved with pure integers:

- `ceil(sum / divisor)`
-  is equivalent to`(sum - 1) / divisor + 1`

Reply:
Also this can be easily unit-tested to safety if we pass the "total_amount" as input instead of passing the mercurial vault :)

Reply:
I think we could create a super-useful utility function here:

`fn compute_precision_loss(total_shares_amount: u64, total_shares_value: u64) -> Result<u64> {}`

then the above could call it like so:

`return compute_precision_loss(total_amount, mercurial_vault_lp_mint_supply)`

That way we would have a re-useable and unit-testable generic functions

Reply:
passing the mercurial vault to this function is avoid the redundancy for getting the `total_amount` with the `get_unlocked_amount` fn with unix timestamp I believe,
but could break this down into two fn and hv the unit-testable `compute_precision_loss` called inside `calculate_possible_lp_token_precision_loss_collateral_value`

Reply:
yep exactly! that's what I meant. the mercurial-specific function calls mercurial.get_unlocked_amount then call the generic function for precision loss :)